### PR TITLE
[Single Span Sampling] Add priority sampling check

### PR DESCRIPTION
**Pull request by marcotc**
Follows up #2103 

This PR adds a new case that was recently clarified from the RFC: non-negative priority sampling values mean that the trace-level sampling chose to reject the trace, and thus Single Span Sampling should apply: https://github.com/DataDog/dd-trace-rb/blob/6eaa041abad95cbf7a88bdb6231be589a48692c8/lib/datadog/tracing/sampling/ext.rb#L11-L22

As a reminder, Single Span Sampling only applies when the trace-level sampling chose to drop the trace.

This PR ensure that Single Span Sampling is applied if:
1. The tracer is dropped inside the tracer (`trace_op.sampled == false`). This means the trace won't be picked up by the transport. Single-sampled spans will be sent individually in the transport in this case.
2. The trace be flushed by the transport (`trace_op.sampled == true`) but, priority sampling is set to a non-positive value.

In case the span is flushed by the transport but there's no priority sampling, Single Span Sampling is not applied, as the trace is not subject to priority sampling, and thus is simply considered as "kept".

### CODE REVIEW

File: lib/datadog/tracing/sampling/span/sampler.rb [hunk 1]
```diff
@@ -23,14 +23,19 @@ def initialize(rules = [])
...
           # If multiple rules match, only the first one is applied.
-          #
+          # @see Datadog::Tracing::Sampling::Ext::Priority Traces with non-positive priorities are considered rejected.
           # @param [Datadog::Tracing::TraceOperation] trace_op trace for the provided span
           # @param [Datadog::Tracing::SpanOperation] span_op Span to apply sampling rules
           # @return [void]
           def sample!(trace_op, span_op)
-            return if trace_op.sampled?
+            if trace_op.sampled? && # Trace will be flushed by the transport.
+               (trace_op.sampling_priority.nil? || trace_op.sampling_priority > 0) # Trace is kept by priority sampling.
```
Comment (ivoanjo) [inline lib/datadog/tracing/sampling/span/sampler.rb:36]: Hmmmmmmmmmmm 🤔. Is it just me or are these concepts really low-level details from the trace operation? For such a simple class that is concerned with span sampling, this is coupling on a lot of really important details that are implemented far away from this file.

TL;DR, have you considered moving this logic inside the trace_op? Something like `trace_op.will_flush?`

**Comment by marcotc** [comment]
Turns out `TraceOperation#sampled?` already encapsulates the logic we need: https://github.com/DataDog/dd-trace-rb/blob/96c395db22c46e08822a2ecc7f8590556952f8f0/lib/datadog/tracing/trace_operation.rb#L116-L118

One thing to notice is that the check for priority sampling is actually slightly different, but `TraceOperation#sampled?` is actually correct (this PR was incorrect).

I'll close this PR as the original behaviour is the desired one.
